### DISCUSSION TITLE
Seeking Advice on Implementing Multi-tenant IAM System

### DISCUSSION THREAD

**Discussion by yordis** [keycloak/keycloak#22090] [Q&A]
I'm currently working on a project where I need an Identity and Access Management (IAM) system that supports multi-tenancy (I think).

The idea is to create a system where users can sign up and become members of distinct organizations.

Users have different roles within each organization, with each role within the organization could carrying specific permissions.

In a situation where you are selling a SaaS service, you create your organization and invite your users into your organization's workspace.

So far, I've explored a solution like Keycloak. It offers robust functionality supporting multi-tenancy, identity management, and access management. However, I still need help deciding the best approach to implement this system, especially configuring each organization as a separate tenant with its users and roles.

I'd love to hear from anyone with experience setting up a multi-tenant IAM system. In particular, I would appreciate insights on:

- Best practices for setting up and configuring multi-tenant IAM systems.
- Recommendations for open-source solutions or other tools that can facilitate the process.
- Any resources, such as books, articles, or online courses, that you found helpful while working on a similar project?

Thank you in advance for your time and insights. I look forward to learning from your experiences!

**Comment by xgp** [comment]
I'd recommend looking at an extension that might be helpful to you, both as a solution to your problem, and to understand some of the tradeoffs of different approaches. (Disclaimer, I'm one of the authors)

https://github.com/p2-inc/keycloak-orgs

**Comment by Mrabdlrhman** [comment]
Hi @yordis,

Did you find a solution to implement an IAM system that support multi-tenancy? I have same case here and I need your advices.

**Reply by yordis** [reply]
https://zitadel.com/

**Reply by Mrabdlrhman** [reply]
Thanks for your prompt response @yordis. Is it open source free or did you use the paid version?

**Reply by yordis** [reply]
100% FOSS, including all the tools needed to run it

**Reply by Mrabdlrhman** [reply]
Thanks for confirming, @yordis . I have a similar situation here and am trying to find a solution. Are you available for a quick call to discuss what you've done on your project? It would give me a clearer picture of where to start, as I have the same case with a SaaS app involving multi-tenancy IAM. If you're not available for a call, no problem at all—any articles or videos that helped you achieve this would be much appreciated. Thank you!

**Reply by yordis** [reply]
Join Zitadel Discord server, they are plenty of active people there helping (including me); Zitadel allows you any combination of B2B, B2C, C2B ... whatever you wish; that is why I moved to it.

In terms of resources, check their YouTube channel
